### PR TITLE
[Sofa.LinearAlgebra] Gather tests

### PR DESCRIPTION
**Pull request by fredroy**
move all tests related to BaseMatrix into Sofa.LinearAlgebra.test, 
and create Sofa.LinearAlgera.Testing (testing toolkit related to Sofa.LinearAlgebra)

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

File: SofaKernel/modules/SofaBaseLinearSolver/CMakeLists.txt [hunk 1]
```diff
@@ -47,10 +47,10 @@ sofa_create_package_with_targets(
...
-# Tests
-# If SOFA_BUILD_TESTS exists and is OFF, then these tests will be auto-disabled
-cmake_dependent_option(SOFABASELINEARSOLVER_BUILD_TESTS "Compile the automatic tests" ON "SOFA_BUILD_TESTS OR NOT DEFINED SOFA_BUILD_TESTS" OFF)
-if(SOFABASELINEARSOLVER_BUILD_TESTS)
-    enable_testing()
-    add_subdirectory(${PROJECT_NAME}_test)
-endif()
+# # Tests
+# # If SOFA_BUILD_TESTS exists and is OFF, then these tests will be auto-disabled
+# cmake_dependent_option(SOFABASELINEARSOLVER_BUILD_TESTS "Compile the automatic tests" ON "SOFA_BUILD_TESTS OR NOT DEFINED SOFA_BUILD_TESTS" OFF)
```
Comment (alxbilger) [inline SofaKernel/modules/SofaBaseLinearSolver/CMakeLists.txt:52]: Commented code

File: SofaKernel/modules/SofaDefaultType/CMakeLists.txt [hunk 2]
```diff
@@ -101,7 +101,7 @@ sofa_create_package_with_targets(
 )
 
 if(SOFA_BUILD_TESTS)
-    add_subdirectory(SofaDefaultType.Testing)
+    # add_subdirectory(SofaDefaultType.Testing)
```
Comment (alxbilger) [inline SofaKernel/modules/SofaDefaultType/CMakeLists.txt]: Commented code

**Review by alxbilger — CHANGES_REQUESTED** [review]
If I understood correctly, there are two testing libraries for `Sofa.LinearAlgebra`:

1. An interface which is only an abstract code waiting to be instantiated with some types. Only header files. This one is Sofa.LinearAlgebra.Testing
2. The actual library which includes some tests in cpp files. This one is Sofa.LinearAlgebra_test. But I don't see any CMakeLists.txt. Did I miss something?

I approve this architecture.
(the names of the libraries are not explicit. It's difficult to understand the difference.)

File: SofaKernel/modules/SofaBaseLinearSolver/CMakeLists.txt (hunk 1 above)
Comment (fredroy) [reply SofaKernel/modules/SofaBaseLinearSolver/CMakeLists.txt:52]: I let commented for the future(?) test in SofaBaseLinearSolver (as the move emptied all the present tests). Would you think we dont care and just delete it ?

File: SofaKernel/modules/SofaDefaultType/CMakeLists.txt (hunk 2 above)
Comment (fredroy) [reply SofaKernel/modules/SofaDefaultType/CMakeLists.txt]: Same

**Comment by fredroy** [comment]
> 
> 
> If I understood correctly, there are two testing libraries for `Sofa.LinearAlgebra`:
> 
>     1. An interface which is only an abstract code waiting to be instantiated with some types. Only header files. This one is Sofa.LinearAlgebra.Testing
> 
>     2. The actual library which includes some tests in cpp files. This one is Sofa.LinearAlgebra_test. But I don't see any CMakeLists.txt. Did I miss something?

I forgot to add the file 🙃
> 
> 
> I approve this architecture. (the names of the libraries are not explicit. It's difficult to understand the difference.)

One naming convention could be: implementations of the tests themselves always end with _test, and the frameworks helping write tests will be suffixed with Testing (i.e sofa::blablabla::testing for Sofa.Blablabla.Testing)
Actually I have more or less convert the existing sofadefaulttype.test and sofadefaulttype.testing for SofaLinearAlgebra

**Review by alxbilger — APPROVED** [review]

**Pull request merged**